simcon_project/conversation_templates/views/template_responses_table.py

```diff
@@ -36,31 +38,54 @@ def get(self, request, pk):
         extra_columns = []  # List of tuples of description and column object to pass to table
         table_data = []  # List of dictionaries to populate table. 1 dictionary = 1 column
 
-        for node in template.template_nodes.all():
-            extra_columns.append((node.description, tables.columns.Column(orderable=False)))
+        for idx, node in enumerate(template.template_nodes.all()):
+            if node.terminal:
+                extra_columns.append((node.description, tables.columns.Column(
+                    orderable=False, attrs={'th': {'class': 'data-column'}},
+                    verbose_name=f'{idx + 1}: {node.description} (Terminal)')))
+            else:
+                extra_columns.append((node.description, tables.columns.Column(
+                    orderable=False, attrs={'th': {'class': 'data-column'}},
+                    verbose_name=f'{idx + 1}: {node.description}')))
+
+        extra_columns.append(("rating", tables.columns.Column(default=False)))
+        extra_columns.append(
+            ("custom_response", tables.columns.BooleanColumn(default=False, verbose_name="Custom End")))
 
         # Populate the table with data for each response
         completed_responses = template.template_responses.exclude(completion_date__isnull=True)
         for response in completed_responses.order_by('-completion_date'):
             column_data = {
-                "first": response.student.first_name,
-                "last": response.student.last_name,
+                "name": f"{response.student.last_name} {response.student.first_name}",
                 "completion_date": response.completion_date,
             }
 
             for node in response.node_responses.all():
-                column_data.update({node.template_node.description: node.transcription})
+                if node.custom_response:
+                    column_data.update({node.template_node.description: node.transcription + ' (Custom Response)'})
+                    column_data.update({"custom_response": True})
+                else:
+                    column_data.update({node.template_node.description: node.transcription})
+                    column_data.update({"custom_response": False})
 
+            column_data.update({"rating": response.self_rating_to_string})
             table_data.append(column_data)
 
+        if 'filter' in request.GET:
+            table_data = filter_search(request, table_data)
+
         if not table_data:
             response_table = None
         else:
             response_table = ResponseTable(data=table_data, extra_columns=extra_columns)
+
+        RequestConfig(request, paginate=False).configure(response_table)
```

Comment:
I get a django error here if the template doesn't have any responses and I go to the excel page.

Reply:
Good catch!